### PR TITLE
Register per span timer/meter for given method

### PR DESCRIPTION
**Pull request by ruhan1**
As described in MMENG-452, it supports:
- Register new field values for the current span
- Register a Timer for a given method, which will accumulate call stats for the current span
- Register a Meter for a given method, which will accumulate marks for the current span

### CODE REVIEW

File: honeycomb/src/main/java/org/commonjava/o11yphant/honeycomb/HoneycombManager.java [hunk 1]
```diff
@@ -228,6 +231,61 @@ public Span getActiveSpan()
...
+        {
+            Timer timer = spanContext.getTimer( name );
+            if ( timer == null )
+            {
+                timer = new Timer();
+                spanContext.putTimer( name, timer );
+            }
+            return timer;
+        }
+        return null;
```
Comment (ligangty) [inline honeycomb/src/main/java/org/commonjava/o11yphant/honeycomb/HoneycombManager.java:247]: One suggestion here: as this method may return null, so maybe we should consider to use java.util.Optional to wrap it, and make all caller easily to check the obj it wrapped, especially for stream programming. This is a best practice from java 8 that let us avoid to return null anymore, which will always bring potential NPE.
So here maybe 
....
   return Optional.of(timer)
}
return Optional.empty()

And from caller:
Optional<Timer> timer = ...
if(timer.isPresent){
timer.get()....
}else{
...
}
So there will be no ugly null checking anymore in the future.

File: honeycomb/src/main/java/org/commonjava/o11yphant/honeycomb/HoneycombManager.java [hunk 2]
```diff
@@ -228,6 +231,61 @@ public Span getActiveSpan()
...
+        SpanContext spanContext = (SpanContext) honeycombContextualizer.extractCurrentContext();
+        if ( spanContext != null )
+        {
+            Timer timer = getSpanTimer( name );
+            if ( timer != null )
+            {
+                return timer.time();
+            }
+        }
+        return null;
```
Comment (ligangty) [inline honeycomb/src/main/java/org/commonjava/o11yphant/honeycomb/HoneycombManager.java:261]: See above.

File: honeycomb/src/main/java/org/commonjava/o11yphant/honeycomb/HoneycombManager.java [hunk 3]
```diff
@@ -228,6 +231,61 @@ public Span getActiveSpan()
...
+        {
+            Meter meter = spanContext.getMeter( name );
+            if ( meter == null )
+            {
+                meter = new Meter();
+                spanContext.putMeter( name, meter );
+            }
+            return meter;
+        }
+        return null;
```
Comment (ligangty) [inline honeycomb/src/main/java/org/commonjava/o11yphant/honeycomb/HoneycombManager.java:277]: See above.

File: honeycomb/src/main/java/org/commonjava/o11yphant/honeycomb/HoneycombContextualizer.java [hunk 4]
```diff
@@ -39,6 +45,8 @@
 
     private static ThreadLocal<Span> SPAN = new ThreadLocal<>();
 
+    private static ThreadLocal<SpanContext> SPAN_CONTEXT = new ThreadLocal<>();
```
Comment (ligangty) [inline honeycomb/src/main/java/org/commonjava/o11yphant/honeycomb/HoneycombContextualizer.java:48]: nitpick: I think the SPAN and SPAN_CONTEXT can be set as final fields

Comment (ruhan1) [reply honeycomb/src/main/java/org/commonjava/o11yphant/honeycomb/HoneycombContextualizer.java]: Ok. Updated.

File: honeycomb/src/main/java/org/commonjava/o11yphant/honeycomb/HoneycombManager.java (hunk 1 above)
Comment (ruhan1) [reply honeycomb/src/main/java/org/commonjava/o11yphant/honeycomb/HoneycombManager.java]: Sounds good. I updated all 3 places.

**Comment by ruhan1** [comment]
This depends on https://github.com/Commonjava/o11yphant/pull/7

File: metrics/core/src/main/java/org/commonjava/o11yphant/metrics/MetricsManager.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import com.codahale.metrics.Timer;
+import com.codahale.metrics.health.HealthCheckRegistry;
+import org.commonjava.cdi.util.weft.ThreadContext;
+import org.commonjava.o11yphant.conf.MetricsConfig;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public interface MetricsManager
```
Comment (jdcasey) [inline metrics/core/src/main/java/org/commonjava/o11yphant/metrics/MetricsManager.java:30]: Is this part of the public API for the library? If so, it seems like we're leaking implementation details (Dropwizard metrics classes).

Comment (ruhan1) [reply metrics/core/src/main/java/org/commonjava/o11yphant/metrics/MetricsManager.java:30]: Yeap. This interface is a draft draft place holder so far. I will refactor it and update you all.

**Pull request merged**